src/test.py

```diff
@@ -50,17 +53,63 @@ def regions(view, result_type):
     return view.get_regions(region_key(view, result_type))
 
 
-def add_markers(view, response):
-    view.run_command("tutkain_clear_test_markers")
+def results(view):
+    """
+    Returns test results persisted in settings.
+    """
+    return view.settings().get(RESULTS_SETTINGS_KEY, [])
+
+
+def unsuccessful(view):
+    """
+    Returns unsuccessful test results, fail and error, persisted in settings.
+    """
+
+    def result_line(result):
+        begin, _ = result["region"]
+
+        return view.rowcol(begin)[0] + 1
+
+    def unsuccessful_type(result):
+        return result["type"] == "fail" or result["type"] == "error"
+
+    l = list(filter(unsuccessful_type, results(view)))
+
+    # Results must be sorted by line.
+    l.sort(key=result_line)
+
+    return l
+
+
+def response_results(view, response):
+    """
+    Returns a dict with "fail", "error" and "pass" keys.
+
+    "fail" and "error" are lists of dict with keys:
+      - "name"
+      - "type"
+      - "region"
+      - "expected"
+      - "actual"
+
+    "pass" is a list of dict with keys:
+      - "name"
+      - "type"
+      - "region"
+    """
 
     results = {"fail": {}, "error": {}, "pass": {}}
 
+    def var_meta_name(result):
```

Comment:
This can just be a variable inside the `for` loop (e.g. on line 110) rather than a function.

Reply:
Nevermind, I misread, my bad. This works.